File: ProtocolAnalysis/src/ProtocolAnalysis/EventHandle/EventDispatchGroup.py
#-*- encoding=utf-8 -*-


from ProtocolAnalysis.DataStruct.MDictionary import MDictionary
from ProtocolAnalysis.Core.AppSysBase import AppSysBase


class EventDispatchGroup(object):
    '''
    classdocs
    '''


    def __init__(self):
        '''
        Constructor
        '''
        self.m_groupID2DispatchDic = MDictionary()
        self.m_bInLoop = False       # 是否是在循环遍历中


    # 添加分发器
    def addEventDispatch(self, groupID, disp):
        if self.m_groupID2DispatchDic.ContainsKey(groupID):
            self.m_groupID2DispatchDic[groupID] = disp


    def addEventHandle(self, groupID, handle):
        self.m_groupID2DispatchDic[groupID].addEventHandle(handle)


    def removeEventHandle(self, groupID, handle):
        if self.m_groupID2DispatchDic.ContainsKey(groupID):
            self.m_groupID2DispatchDic[groupID].removeEventHandle(handle)
        else:
            AppSysBase.instance().m_logSys.log("Event Dispatch Group not exist")


    def dispatchEvent(self, groupID, dispatchObject):
        self.m_bInLoop = True
        if self.m_groupID2DispatchDic.ContainsKey(groupID):
            self.m_groupID2DispatchDic[groupID].dispatchEvent(dispatchObject)
        else:
            AppSysBase.instance().m_logSys.log("Event Dispatch Group not exist")
        self.m_bInLoop = False;


    def clearAllEventHandle(self):
        if not self.m_bInLoop:
            for dispatch in self.m_groupID2DispatchDic.Values():
                dispatch.clearEventHandle()

            self.m_groupID2DispatchDic.Clear()
        else:
            AppSysBase.instance().m_logSys.log("looping cannot delete element");


    def clearGroupEventHandle(self, groupID):
        if not self.m_bInLoop:
            if self.m_groupID2DispatchDic.ContainsKey(groupID):
                self.m_groupID2DispatchDic[groupID].clearEventHandle()
                self.m_groupID2DispatchDic.Remove(groupID)
            else:
                AppSysBase.instance().m_logSys.log("Event Dispatch Group not exist");
        else:
            AppSysBase.instance().m_logSys.log("looping cannot delete element");

File: LuaLibs/src/Libs/EventHandle/EventDispatchGroup.lua
local EventDispatchGroup = GlobalNS.Class()
local M = EventDispatchGroup

function M:ctor()
    self.m_groupID2DispatchDic = GlobalNS.Dictionary:new()
    self.m_bInLoop = false;
end

-- 添加分发器
function M:addEventDispatch(groupID, disp)
    if not self.m_groupID2DispatchDic.ContainsKey(groupID) then
        self.m_groupID2DispatchDic.Add(groupID, disp);
    end
end

function addEventHandle(groupID, handle)
    self.m_groupID2DispatchDic.value(groupID).addEventHandle(handle)
end

function removeEventHandle(groupID, handle)
    if self.m_groupID2DispatchDic.ContainsKey(groupID) then
        self.m_groupID2DispatchDic.value(groupID).removeEventHandle(handle)
    else
        -- 日志
    end
end

function dispatchEvent(groupID,  dispatchObject)
    self.m_bInLoop = true;
    if self.m_groupID2DispatchDic.ContainsKey(groupID) then
        self.m_groupID2DispatchDic.value(groupID).dispatchEvent(dispatchObject)
    else
        -- 日志
    end
    self.m_bInLoop = false;
end

function M:clearAllEventHandle()
    if not m_bInLoop then
        for _, dispatch in pairs(m_groupID2DispatchDic) do
            dispatch.clearEventHandle()
        end

        m_groupID2DispatchDic.Clear()
    else
        -- 日志
    end
end

function clearGroupEventHandle(groupID)
    if not self.m_bInLoop then
        if self.m_groupID2DispatchDic.ContainsKey(groupID) then
            self.m_groupID2DispatchDic.value(groupID).clearEventHandle();
            self.m_groupID2DispatchDic.Remove(groupID);
        else
            -- 日志
        end
    else
        -- 日志
    end
end

return M